Display active runtime metadata in console pane

dhruvisompura opened the pull request:
Addresses #5999 

Adds a new feature flag called `positron.multipleConsoleSessions`. When the feature flag is on users can now view details for an active session by clicking on the new info icon button that lives in the console pane action bar. A modal popup will render, containing the details of the session, such as name, state, id, interpreter path, and environment management tool.

The popup provides a link to the session logs in the output pane. When clicked, the user is taken to the output pane and shown the kernel channel for their interpreter.

This is a first pass for the info panel - additional information may be added based off feedback! Follow up work to add additional links to other channels is covered in #6149.

### Release Notes

#### New Features

- N/A

#### Bug Fixes

- N/A


### QA Notes
- [ ] Verify the info button is only shown when the feature flag is on and there is an active session for the console instance
  - Note: An issue has been noted where the popup is not rendered if the console pane height is smaller than the height of the popup (#3061)
- [ ] Verify the session status shown in the popup updates in real time

### Screenshots

**Feature Flag**
<img width="1272" alt="Screenshot 2025-01-29 at 2 19 30 PM" src="https://github.com/user-attachments/assets/84154034-f7a7-4696-b016-61c4aa3c9370" />

**Popup**
<img width="1330" alt="Screenshot 2025-01-29 at 4 38 16 PM" src="https://github.com/user-attachments/assets/fc819ef6-ad75-4e59-9217-47b97c965b41" />

**Real-Time Session State Updates**

https://github.com/user-attachments/assets/16b0b81f-7b48-43d0-9f54-047e6e2af4be

dhruvisompura committed:
Add feature flag for multi console sessions

dhruvisompura committed:
Add console session info button

dhruvisompura committed:
Display basic session data

dhruvisompura committed:
Add action to open output channel

dhruvisompura committed:
Display session state

dhruvisompura committed:
Update colors and markup

softwarenerd commented on src/vs/workbench/contrib/positronConsole/browser/components/consoleInstanceInfoButton.css:
```diff
@@ -0,0 +1,32 @@
+}
+
+.console-instance-info .content .line {
+	margin: 8px 0px;
```
I'd use `margin: 8px 0;`. Never `0px`.

softwarenerd commented on src/vs/workbench/contrib/positronConsole/browser/components/consoleInstanceInfoButton.css:
```diff
@@ -0,0 +1,32 @@
+	cursor: pointer;
+	font-size: 12px;
+	line-height: 22px;
+	padding: 0px 8px;
```
I'd use `padding: 0 8px;`. Never 0px.

softwarenerd commented on src/vs/workbench/contrib/positronConsole/browser/components/consoleInstanceInfoButton.tsx:
```diff
@@ -0,0 +1,125 @@
+			setSessionState(state);
+		}));
+		return () => disposableStore.dispose();
+	});
```
This `useEffect` will run on every render. It should have an empty dependencies array so that it runs only once.

```
// Main useEffect hook.
	useEffect(() => {
		const disposableStore = new DisposableStore();
		disposableStore.add(props.session.onDidChangeRuntimeState(state => {
			setSessionState(state);
		}));
		return () => disposableStore.dispose();
	}, []);
```

softwarenerd commented on src/vs/workbench/contrib/positronConsole/browser/components/consoleInstanceInfoButton.tsx:
```diff
@@ -0,0 +1,125 @@
+
+	// Render.
+	return (
+		<PositronModalPopup
```
`Session ID:`, `State:`, etc. below should be localized.

softwarenerd commented on src/vs/workbench/services/runtimeSession/common/positronMultipleConsoleSessionsFeatureFlag.ts:
```diff
@@ -0,0 +1,51 @@
+	ConfigurationScope,
+	Extensions,
+	IConfigurationRegistry,
+} from '../../../../platform/configuration/common/configurationRegistry.js';
```
Nit: We don't typically use multiple lines for `import` statements.

juliasilge reviewed (dismissed):
Working well for me!

![Screenshot 2025-01-30 at 4 04 03 PM](https://github.com/user-attachments/assets/3c264bbc-bc49-4b64-867e-708c8bad4de4)

Can we open a followup issue to re-examine the treatment of the top action bar of the console at narrower widths? Especially after the coming major UI changes (probably not worth it before). See for example how when I choose the "side-by-side" layout at a moderate app size, I get some not great behavior:

https://github.com/user-attachments/assets/1337d0fd-4e7b-4248-a2d0-c299dbc25b65

softwarenerd reviewed (dismissed):
Works for me, too!

dhruvisompura committed:
Clean up CSS syntax

dhruvisompura committed:
Add localization

dhruvisompura committed:
Condense multi-line import to single line

dhruvisompura committed:
Remove checks for undefined session

No need to check if a session is undefined because we are
guaranteed to get a session

dhruvisompura committed:
Prevent useEffect from running on every render

softwarenerd reviewed (dismissed):
LG!

sharon-wang commented on src/vs/workbench/contrib/positronConsole/browser/components/consoleInstanceInfoButton.tsx:
```diff
@@ -0,0 +1,135 @@
+			popupPosition='auto'
+			width={400}
+		>
+			<div className='console-instance-info'>
```
A idea for the future -- thoughts on making the text copy-able? In particular, it would be great if the interpreter path was copy-able.

<img width="408" alt="image" src="https://github.com/user-attachments/assets/e750ac08-060e-4dff-9c38-935bf83e7ae9" />

dhruvisompura replied:
Yes actually! This is something I noticed and made a note about in the epic for future improvements!

sharon-wang reviewed (dismissed):
Also working well for me in dev Server Web and Desktop on Mac 👌

LGTM!

sharon-wang commented on src/vs/workbench/contrib/positronConsole/browser/components/consoleInstanceInfoButton.tsx:
```diff
@@ -0,0 +1,135 @@
+}
+
+const ConsoleInstanceInfoModalPopup = (props: ConsoleInstanceInfoModalPopupProps) => {
+	const [sessionState, setSessionState] = useState(props.session.getRuntimeState());
```
Throwing back to a [React tip from Nick](https://github.com/posit-dev/positron/pull/2899#pullrequestreview-2028647997) -- we can avoid calling `getRuntimeState()` on every re-render by wrapping the function call in an arrow function:

```suggestion
	const [sessionState, setSessionState] = useState(() => props.session.getRuntimeState());
```

(re-commented as the original suggestion looked to be on the wrong line, not sure how that happened!)

dhruvisompura replied:
Thank you for pointing that out! I've gone ahead and made this change. I also updated the debug session component to have the same improvements.

dhruvisompura committed:
Reduce unnecessary computations per render cycle

sharon-wang reviewed (approved):
:shipit:!
